### DISCUSSION TITLE
How to get stats (lines +/-) of diff between index and workdir ?

### DISCUSSION THREAD

**Discussion by Canop** [GitoxideLabs/gitoxide#1366] [Q&A]
I'm looking for the equivalent of [diff_index_to_workdir](https://docs.rs/git2/latest/git2/struct.Repository.html#method.diff_index_to_workdir), which gives the number of lines added/removed in work directory compared to index.
Is it possible to get it or compute it with gix, starting from a Repository ?

**Comment by Byron** [comment]
There is a nice example for how to do this with granularity of *files* [here in onefetch](https://github.com/o2sh/onefetch/blob/ed99fe2c89bfba52a12dcbf4d03264d2ae9c714d/src/info/pending.rs#L21).

In order to bring that down to *lines* granularity one will have to make use of a [diff_resource_cache()](https://docs.rs/gix/latest/gix/struct.Repository.html#method.diff_resource_cache) which unfortunately right now isn't tightly integrated into the index-worktree diff objects yet so it would require some fiddling. Examples are plentiful in the codebase though if you are interested to give it a try nonetheless.

**Reply by blinxen** [reply]
Coming back to this, did the API evolve in the past year? I tried to make this work but it did not seem to work *easily*. My goal is to remove `libgit2` from `bat`.

Just to clarify what I want to achieve:

* I have a git repository that contains some commits
* I edit a file that is tracked by git
* Now I want to create a diff that shows me how I can get from the `HEAD` commit to the modified file

As far as I can see, the git index does not have to be involved in this whole operation (or does it?). Is that currently possible with this library?

**Reply by Byron** [reply]
It will take some getting used to in comparison to `git2`, but it can absolutely be done today.

With a [`diff_resource_cache()`](https://docs.rs/gix/latest/gix/struct.Repository.html#method.diff_resource_cache) one can set both sides up for diffing, one comes from the blob in the index, the other comes from the worktree directly. Then one can prepare are diff from which one can finally perform a diff using `gix::diff`. It's all fairly low-level and one will have to deal with hunks generated from the diff directly.

Feel free to ping me in the PR if you decide to give `gitoxide` a try.

**Reply by blinxen** [reply]
Hello again! I got busy at work and did not have time these past weeks. Now I wanted to give this a shot. What I have now is the following:

```rust
let filename_bstr = BStr::new(filename.to_str()?.as_bytes());
let repository = gix::discover(filename.canonicalize().ok()?.parent()?).ok()?;
let mut cache = repository.diff_resource_cache(Mode::ToGit, Default::default()).ok()?;
cache.set_resource(
    ?????,
    EntryKind::Blob,
    &filename_bstr,
    ResourceKind::OldOrSource,
    &repository
).ok()?;
cache.set_resource(
    ObjectId::null(Kind::Sha1),
    EntryKind::Blob,
    &filename_bstr,
    ResourceKind::NewOrDestination,
    &repository
).ok()?;
let interned_input = cache.prepare_diff().ok()?.interned_input();
let diff = gix::diff::blob::diff(
    Algorithm::Myers,
    &interned_input,
    // struct implementing the Sink trait
    DiffStepper
);
```

With this I can get something working but it is not what I expect. I think my problem is that I set the wrong `ObjectId`s for the resources. For example if I set the old source as `repository.head_id().ok()?.into()`, I can see that I am diffing a commit blob with an empty file. That is very wrong :D. What am I doing wrong?

**Reply by Byron** [reply]
The ObjectID provided as `OldOrSource` must be the one of a blob, whereas `repository.head_id()` is a commit. It's interesting it doesn't error, but instead uses an empty blob.
Oh, now I see that it does use the commit data itself. An empty blob may be used if the file doesn't exist (and the ObjectID is null).

**Reply by blinxen** [reply]
I assumed that I had to specify the commit and `gitoxide` would magically retrieve the correct file since I also pass the file path  to the `set_resource`. Now it makes a bit more sense but to be honest I am still pretty lost on how to get the object ID of the file from both index and worktree. I can't seem to find any examples on how to work with the index / worktree. Also, why would files that are tracked by git but changed on disk have an object id?

I appreciate the time you take to answer me, I know you are quite busy ;D.

**Reply by Byron** [reply]
I think the documentation of the function explains it sufficiently well, but if not you are welcome to help make it more clear.
In short, a null object id should be provided if the file should be read from the worktree. Otherwise, it will always be read from the object database.

**Reply by blinxen** [reply]
After trying out a couple of things, I can see that the documentation was indeed sufficient. My problem here was that when building the cache, I used the default values for the `WorktreeRoots` struct, which were `None`. So when trying to access the file in the worktree, `gix` would always use an empty blob. Thanks for the help!

**Reply by Byron** [reply]
Thanks for sharing, maybe this is something that should be fixed, as I have trouble seeing a use-case where the ObjectId would be null, and no worktree root is set, which always should mean that it tries to read from disk.
And it's fixed via https://github.com/GitoxideLabs/gitoxide/pull/1898 .

**Comment by Canop** [comment]
I understand you refer for example to this : https://github.com/Byron/gitoxide/blob/c4e0a82ec9948ed0d9876a50b156d1ba30e2ad77/gix/tests/object/tree/diff.rs#L11-L53 ?

Does that work for comparing the workdir to the index ?

I'll try. Thanks.

**Reply by Byron** [reply]
Apologies, but this code would compare two trees with each other, and it has nothing to do with the index file or the work-tree.

**Comment by Canop** [comment]
BTW, the reason I ask is I'm trying to replace git2 with gix in broot: https://dystroy.org/broot/#check-git-statuses
